README.md
USER GUIDE TO THE APOLLO ATMOSPHERE RETRIEVAL CODE

APOLLO v0.10.2 Beta
30 December 2019

Public Beta Release: 17 July 2020

NOTE: this public beta is generally stable for the cases described in 
Section 2, but development is ongoing. Expect a major overhaul with 
v0.11 in Q3 2020.

Molecular cross section files are not included with this release due to 
memory constraints. Please contact the developer at 
alex.r.howe@nasa.gov for more information, or build your own based on 
the format described in Section 8.

Alex R. Howe

Included files:
Apollo.tar
	README.txt
	maketar.sh
	APOLLO source code
	Makespectrum source code
	MakeHaze source code
	Plot Apollo source code
	Plot Converge source code
	Plot Opac source code
	JWST filter list
	JWST filter functions
	Example files

CONTENTS
1.  Description
2.  Functionality
3.  Installation
4.  Running
5.  Input File Format: Settings
6.  Input File Format: Parameters
7.  Spectrum File Format
8.  Molecular Cross Section Table Format
9.  Haze Table Format
10. Output File Formats

1. DESCRIPTION

APOLLO is an atmosphere code for extrasolar planets designed to fit
forward models to both transit and emission spectra and retrieve atmospheric
properties. APOLLO uses an MCMC algorithm to find the best fit to a spectrum
for a given parameter set. The code is designed to be modular enough to have
many options that may be turned on and off, including the type of
observations, a flexible molecular composition, multiple cloud prescriptions,
multiple temperature-pressure profile prescriptions, multiple priors, and
continuum normalization. It uses cgs units unless otherwise specified.

APOLLO is written in C++ and Python 2.7 using the Cython extension. (Python 3
is in the works.) Anaconda is the recommended build of Python given that scipy
is required. The emcee and corner packages are needed in addition to Anaconda.
The original build environment includes:

Python 2.7.13
Anaconda 4.4.0 for Python 2
emcee 2.2.1
corner 2.0.1
Cython 0.25.2
GCC 4.9.4

NOTE: the version of emcee is especially important. Emcee version 3 makes
significant changes an is not fully backward-compatible. In particular, the
MPI Pool functionality is forked to another module, schwimmbad, and the sample
chain is handled with a new Backends class.

The build environment needed to run APOLLO on a computing cluster will vary
with the particular cluster. You may need to load Anaconda and the C++
compiler, and possibly create a Python virtual environment. Contact your local
support team for assistance.

NOTE for Discover: to convert from the local version that uses Emcee 2
and the on-cluster MPI version that uses Emcee 3, do the following:
1. In setup.py, change "g++-mp-9" to "g++" and "gcc-mp-9" to "gcc".
2. In Apollo.py, change:
    import corner
To:
    #import corner
    import matplotlib
    matplotlib.use('pdf')
3. In Apollo.py, change:
    from emcee.utils import MPIPool
To:
    from schwimmbad import MPIPool
4. In Apollo.py, before setting up the sampler, add:
    reader = emcee.backends.Backend()
5. In the EnsembleSampler constructor, add the parameter "backend=reader".
6. In the sample iteration loop, add at the beginning:
    coords = reader.get_last_sample().coords
    blobs = reader.get_blobs()[i]
7. In the sample iteration loop, change "sample[0]" to "coords" and "sample[3]"
    to "blobs" at each instance.
8. Comment out the entire plotting section.

2. FUNCTIONALITY

APOLLO version 0.10 is verified to be accurate and stable with the
one-stream radiative transfer algorithm for transit, secondary eclipse, and
direct imaging; for layered and 5-parameter T-P profiles; for cloud-free
models, an opaque cloud deck, a cloud slab of uniform particle size and
density, and a cloud of gaussian particle density and uniform particle size.

The two-stream radiative transfer algorithm is verified only for cloud-free and
opaque cloud deck eclipse spectra.

APOLLO is bundled with the following programs:

Apollo Makespectrum program computes spectra of forward models complete with
noise levels for a model JWST pipeline. Makespectrum uses the same input file
format as APOLLO, although it uses a slightly different set of parameters. It
generates spectrum files that can be used as input "observations" in APOLLO.
Makespectrum currently produces only JWST spectroscopic and photometric modes.
However, it does output the full-resolution spectrum to compute noise levels
for model pipelines for other observatories.

Plot Apollo, a standalone version of the plotting routines for APOLLO and
Makespectrum.

Plot Converge plots the motion of the walkers in each dimension to the screen.
This can be useful to troubleshoot if the walking is malfunctioning.

Plot Opac plots two sets of cross sections from different tables to the screen
for a particular pressure and temperature. This is meant to compare similar
tables to check for any any differences.

MakeHaze computes new haze opacity tables using Mie scattering from a table of
indices of refraction for the haze material. Note that MakeHaze is a C++
program, unlike the others.

3. INSTALLATION

To install APOLLO, unpack the Apollo.tar tarfile, and compile the code IN THE
SRC DIRECTORY using the Cython setup file:

python setup.py build_ext --inplace

This script contains all of the setup needed to run both APOLLO and
Makespectrum. APOLLO is also bundled with tables of molecular and haze
particle cross sections for NIR and MIR observations, which are also necessary
to run the code. However, these tables total 10.6 GB, so due to their size,
they are available only via USB drive or FTP.

NOTE: You will need to set the C and C++ compiler names in setup.py to the
compilers used by your system.

By default, these tables are read from a directory named Opacities in the
parent directory to your work directory. This allows multiple builds on the
code to be run in different directories. However, the memory allocation of a
cluster is often such that you may need to store the opacity tables in an
archival or scratch directory. The filepath to this directory is specified with
the Opacities parameter in the input file.

Use the included Makefile to build the MakeHaze program.

4. RUNNING

APOLLO is normally called with a single command line argument, which is the
name of the input file. If no input file is specified, it will default to the
bundled example file, example.resolved.dat. If a second command line argument,
"Serial", is added, it will override parallel operation and initiate a
minimal-length serial run for testing purposes. (This should take less than 10
minutes on a modern processor.) APOLLO may also be run in serial mode if
parallel mode is turned off in the input file.

In serial model, APOLLO may be run as a regular Python program:

python Apollo.py <input file> Serial

The default mode of APOLLO is parallel processing. On Flux, this can be done
using a PBS script incorporating the following command:

mpirun python Apollo.py <input file> >> <scratch directory>/Apollo.log

Sending the log output to a scratch directory is often needed on a cluster to
reduce memory usage in the work directory. If it is not included, the log file
will be sent to the work directory. Other MPI arguments may also be needed,
depending on the cluster.

The example files include the parameters for GJ 570D for resolved spectra, and
for HD 189733b for transit and eclipse spectra.

NOTE: Different clusters may require different settings to run APOLLO
efficiently. Examples:

Discover at NCCS requires nodes=8 (28 cores per node), ntasks=216, and
cpus-per-task=1.

Great Lakes at U of M requires... (36 cores per node)

The old Flux cluster at U of M (which used PBS) requires nodes=14 and ppn=16
(16 cores per node).

Apollo Makespectrum is run as a regular Python program and does not need MPI:

python Apollo.makespectrum.py <input file>

Plot Apollo is run from the command line as Plot.Apollo.py. It can replicate
the output plots for both APOLLO and Makespectrum. Plot Apollo takes 4 command
line arguments:
1. Input file. This is a sample file or a spectrum file.
2. Object name. This will determine the output file names.
3. Data type. "Samples" for a samples file, which will create the APOLLO plots.
"Spectrum" for a spectrum file, which will create the Makespectrum plot.
4. Reference spectrum. Required for spectrum files only. The input (observed)
spectrum to plot against the spectral fit. Residuals will be calculated by
binning the spectral fit to the reference spectrum.
4. Confidence interval. Optional for samples files. Sets the bounds on the
histograms such that the specified fraction of samples fall within the
waterfall plots. Default: 0.99

Plot Converge is run from the command line as Plot.Converge.py. It plots the
motion of all of the walkers over a single dimension in each frame and will
show if APOLLO has converged. It takes a samples file name as a command line
argument. By default, it displays the plots in a window instead of saving them.

Plot Opac is run from the command line as Plot.Opac.py. It plots the molecular
cross sections from two specified tables at a particular pressure and
temperature. By default, it displays the plot in a window instead of saving it.
Plot Opac takes 4 command line arguments:
1. First cross section file name.
2. Second cross section file name.
3. Pressure in mbar.
4. Temperature in kelvin.
NOTE: Plot Opac rounds down to the nearest pressure-temperature point instead
of interpolating. This is because it is meant to compare the actual values in
the tables for similar tables, and interpolation could mask the actual
differences.

MakeHaze is run from the command line as ./MakeHaze. It takes 3 optional
command line arguments:
1. Opacities directory for both input and output files. Default: ../Opacities
2. Input file name, a table of real and complex indices of refraction versus
wavelength. Default: blank.r.dat, a table with all indices of refraction set
to 1, representing no refraction.
3. Output file name. Default: haze.table.dat

Note that MakeHaze.cpp should be able to compute >100 wavelengths x 161
particle sizes per minute. If it is taking significantly longer, there may
be a problem with the indices file.

5. INPUT FILE FORMAT: SETTINGS

The input file for APOLLO and Makespectrum consists of two sections: a list of
settings for a particular configuration of the code, and a list of parameters
for the forward model to be fit. The settings may be listed in any order, but
they all must come before the parameters. The parameters must be grouped into
blocks with specific labels.

Defaults are set so that any parameters may be omitted from the input file. The
header line for the parameter section must be included, but the code will
execute without any other lines present.

The input files are designed so that they can be used interchangeably with
APOLLO and Makespectrum, so some variables that are used for one are dummies in
the other.

The SETTINGS block of the input file may contain any of the following settings.
(Settings not included in this list will be ignored.)

Mode, 1 field:
1. Type of spectrum. "Resolved" for resolved objects in emission. "Transit"
for transit spectra. "Eclipse" for secondary eclipse. Default: "Resolved"

Parallel, 1 field:
1. Mode of operation. "Parallel" for parallel operation using the MPI Pool.
"Serial" for serial operation from the command line.
NOTE: The "Serial" command line argument overrides this setting.

Object, 1 field:
1. The name of the target object. This field may be anything and is used only
to set the name of the output file. Default: "example"

Star, 3 fields:
1. Stellar Temperature in kelvin. Default: 5770 (Solar temperature)
2. Stellar radius in Solar radii. Default: 1.0
3. Semi-major axis of the planet in AU. Default: 1.0
NOTE: These parameters are not used in the "Resolved" mode.

Location, 3 fields:
1. Distance in parsecs. Default: 10.0
2. Right ascension in decimal degrees. Default: 0.0
3. Declination in decimal degrees. Default: 0.0

Data, 3 fields:
1. Spectrum file name. This file contains the observations to be fit. Default:
"example.obs.dat" (the bundled example file)
2. Binning factor observations. This is an integer factor by which to bin down
the observations. This may be useful because it is recommended for the forward
model to be 5-10 times higher resolution than the observations for the most
reliable results. However, if the spectrum has gaps or varies significantly in
resolution, it should be set to 1. Default: 1
3. Polynomial fitting flag. If the codeword "Polyfit" is included, the code
will normalize a subset of observations to a polynomial fit based on another
set of observations. The boundary between the two sets is set where the
wavenumber of the observations in the input spectrum increases.

Star_Spec, 1 field:
1. Stellar spectrum file name. Used only in Makespectrum. Defines a stellar
spectrum to use in the noise model in place of a blackbody. If no file is
specified, it will default to a blackbody spectrum.

Degrade, 1 field:
1. Degradation factor. This is an integer factor by which to degrade the
spectral resolution of the forward model. Degrading the spectrum speeds up
the code proportionally, although the lower resolution may reduce the
reliability of the code. Default: 3
NOTE: This setting is hard-coded to 1 in Makespectrum because this program is
intended to produce a high-resolution model.

N_Walkers, 1 field:
1. Number of walkers to be initialized by emcee. Default: 8*N, where N is the
number of parameters.
NOTE: if this number is set too low, it will reset to the minimum allowed value
of 2*N+2.
NOTE: N_Walkers must be even. If it is odd, the code will add 1.
NOTE: The "Serial" command line argument overrides this setting and sets it to
the minimum value.

N_Steps, 1 field:
1. The number of steps for the MCMC algorithm to take. This must be enough for
the walkers to converge and may require adjustment for any given object.
Default: 30000
NOTE: The "Serial" command line argument overrides this setting and sets nsteps
to 2.
NOTE: autocorrelation convergence checking is not yet implemented.

Pressure, 2 fields:
1. Minimum log-pressure in bars over which to compute the radiative transfer.
Default: -3.0
2. Maximum log-pressure in bars over which to compute the radiative transfer.
Default: 2.5

Vres, 1 field:
1. Vertical resolution: number of layers to use for the radiative transfer
calculation. Default: 71
NOTE: This applies to both the "Layers" and "Parametric" T-P profiles.

Streams, 1 field:
1. Which radiative transfer function to use. 1 for 1-stream, 2 for 2-stream.
Default: 1
NOTE: the 2-stream algorithm is much slower than the 1-stream algorithm, by as
much as 2 orders of magnitude. It is currently implemented only for cloudless
atmospheres and opaque cloud decks in emission.

Prior, 1 field:
1. Type of prior to be used. Currently supported priors are "Uniform" and
"Normal". Default: "Uniform"

Gray, 2 fields:
1. Boolean switch to use a gray atmosphere model. Default: False
2. Temperature of the gray atmosphere model. Default: 1500 K.

Output, 3 fields:
1. Output directory for the samples file. This may need to be a scratch or
archival directory on a cluster. This field is included only for the samples
file in APOLLO because of the large sizes of these files. All other outputs
automatically go to one of the local subdirectories included in Apollo.tar,
either "modelspectra" or "plots". Default: "samples"
2. Short filename flag. If the codeword "Short" is included, the code will
create output files with short names incorporating only the object name rather
than the default names, which incorporate other properties of the particular
calculation.
3. Print full output flag. If the codeword "Full" is included, the code will
output the full sample array to the samples file instead of the last 10% of the
array. This allows analysis of the burn-in phase of the run.
NOTE: The "Serial" command line argument overrides this setting and sets it to
True.
NOTE: The order of "Short" and "Full" may be reversed for APOLLO.
NOTE: The "Short" flag may be in any position for Makespectrum.

Opacities, 1 field:
1. Opacity files directory. This will often need to be a scratch or archival
directory on a cluster. Default: "../Opacities"

Tables, 2 fields:
1. Set of opacity tables to use for computing the spectrum. For example,
listing "nir" will use files of the form "mol.nir.dat".
NOTE: The default for the opacity table selection is for the code to
automatically select "nir", "mir", or "wide" depending on the wavelengths
covered by the observed spectrum.
2. Set of opacity tables to use for computing the effective temperature. The
effective temperature must be computed over a wide wavelength range, so a much
lower spectral resolution is needed to do it efficiently. Default: "lores"

Currently included tables:
"nir": 0.6-5.0 microns, R=10,000.
"mir": 5.0-30.0 microns, R=10,000.
"wide": 0.6-30.0 microns, R=10,000.
"lores": 0.6-30.0 microns, R=200.

Output_Mode, 1 field:
1. Output spectroscopic or photometric mode. This is used only in Makespectrum.
If one of JWST's spectroscopic modes is specified, Makespectrum will create a
spectrum with noise using a pipeline for that mode in addition to the
full-resolution spectrum. If a file containing a filter throughput function is
specified, Makespectrum will compute the weighted mean flux through the filter.
These filter functions must have two columns:
1. Wavenumber in cm^-1
2. Fractional throughput.
By default, this parameter is not used.

Supported JWST modes:
NIRISS
NIRCam
NIRSpec1H
NIRSpec2H
NIRSpec3H
NIRSpec4H
NIRSpec4M
Prism

6. INPUT FILE FORMAT: PARAMETERS

The PARAMETERS block of the input file MUST begin with a header beginning with
the string "Parameter". This header shows the format of the parameters section:

"Parameter   Initial   Mu    Sigma    Min     Max"

The model parameters are listed in blocks with headers to separate them in the
code's handling. There are 5 blocks currently implemented, which may be in any
order, and any individual block may be omitted. However, if a block is
included, general all of the parameters for that block MUST be included. Each
parameter is listed with 6 fields:
1. Parameter name.
2. Initial guess.
3. Mean of the prior. (Only used for normal priors.)
4. Standard deviation of the prior. (Only used for normal priors.)
5. Minimum cutoff of the prior.
6. Maximum cutoff of the prior.

The implemented values of each block are listed below.

Basic
Bulk parameters of the planet. Currently supported parameters:
Rad: Radius in Earth radii
RtoD: Log of radius-to-distance ratio
RtoD2U: Log of squared radius-to-distance ratio
Default radius: 1 Jupiter radius in linear space
Log(g): log(g) in cm/s^2. Default: 4.1
Cloud_Base: log-pressure in bars at the cloud deck that serves as the base of
the atmosphere for the forward model. Default: 2.5
P_cl: alternate name for Cloud_Base.

NOTE: If the radius is not specified, the default is set to 1 Jupiter radius.
However, the code will automatically normalize the spectrum to the total flux
of the observations.
NOTE: APOLLO includes an internal prior that the mass must be less than 80
Jupiter masses. It is important to set the initial radius and gravity so that
the mass falls below this limit.
NOTE: Cloud_Base/P_cl allows the inclusion of an opaque cloud deck in addition
to another cloud model. (This also works to represent the limits imposed by
refraction on probing transit spectra.)

Gases, 1 field:
1. Filler gas. This gas will comprise  the balance of the atmosphere after the
listed gases in the model are accounted for, subtracting their abundances
from 1. Any of the supported molecules can be used as the filler gas.
Default: "h2-only".

List of molecules included in the model in log-abundance. Currently supported
molecules:
h2 (H2 + He)
h2only
he
h-
h2o
ch4
co
co2
nh3
h2s
Burrows_alk (Burrows Alkali Metals)
Lupu_alk (Lupu Alkali Metals)
crh
feh
tio
vo
hcn
n2
ph3

NOTE: The Lupu_alk option is currently disabled because of difficulties in
parsing the tables for use by APOLLO.

Atm, 2 fields:
1. Type of T-P profile for the atmosphere. "Layers" for an arbitrary number of
temperature points spaced evenly in log-pressure space. "Parametric" for a
5-parameter analytic T-P profile taken from Line et al. (2012).
2. Override vertical resolution flag. If the code word "Verbatim" is included,
the code will use a layered profile directly from the input file instead of the
standard (usually higher) vertical resolution.
NOTE: If the T-P profile is omitted entirely, the code will default to an
isothermal atmosphere with a temperature of 1500 K.

Parameters for the "Layers" profile:
T#: an arbitrary number of temperature points in kelvin. 15 is recommended.
gamma: a smoothing parameter that penalizes wide swings in the T-P profile in
the likelihood function. If included, the program will automatically implement
a smoothed T-P profile. Defined using an Inverse Gamma Function as in
Line et al. (2015).
NOTE: Although the peak in the distribution for gamma is set to 5.e-5, the
upper bound must be set to 1000 because this is the approximate point where the
roughness penalty approaches its asymptotic value.
NOTE: It may be necessary to set the initial guess for gamma high, around ~10,
to avoid the code getting stuck in a local minimum with too much smoothing.

NOTE: A single T# point will result in an isothermal atmosphere.
NOTE: The default cross section tables have boundaries of 75 K and 4000 K,
which are the default limits on the priors.

Parameters for the "Parametric" profile:
Tint: Effective temperature due to internal heat only.
log(kappaIR): Mean opacity in the infrared.
gamma1: gamma1 parameter from Line et al. (2012).
gamma2: gamma2 parameter from Line et al. (2012).
alpha: alpha parameter from Line et al. (2012).

NOTE: If "Parametric" is specified, and the wrong number of parameters are
listed, the code will halt with an error message.

Clouds, 2 fields:
1. Cloud model number.
2. Haze type (optional).

Currently supported cloud (aerosol) models:
0: No clouds (0 parameters).
1: Opaque cloud deck (1 parameter).
2: Slab cloud profile with uniform particle size (4 parameters).
3: Gaussian cloud profile with uniform particle size (4 parameters).

Model 1:
P_cl: log-pressure in bars at the cloud deck that serves as the base of
the atmosphere for the forward model.


ChangeLog.txt
<<<<<<< Updated upstream
APOLLO TO-DO LIST

Bugs and Bug Detection:
2. Check that the metallicity, etc. are being computed correctly when there
   are fixed parameters.
4. Give absorption and scattering cross sections separate table sizes.

Pending Updates:
5. Implement Arthur's Planet_auto_4-models_exponential.cpp.
6. Implement Arthur's changes to output and plot both median and MLE values.
7. Implement and example Arthur's addition of cloud parameters and parametric
   T-P profiles to Plot Apollo.
8. Implement Arthur's contribution functions (once completed).

Cosmetic:
9. Reorganize the header file to handle public and private correctly.
10. Replace all example output files with up to date versions.
11. Update the inline documentation and README file. (Include any NOTES from
    the Change Log.)
12. Revise the code for clarity: less confusing names for some parameters,
    correct labeling of "cross section" vs. "opacity" and "aerosol" vs. "cloud"
    or "haze".
13. Create or modify an example file for Ensemble mode.

Aerosol and T-P Models:
14. Combine layer-by-layer and (multiple?) parametric T-P profile generation
    functions into a single pipeline.
15. Combine the various aerosol models into a single pipeline (with clear names
    for each model).
16. Add more cloud models, especially particle size distributions.
17. Set the cloud priors so that the cloud base cannot be deeper than the
    maximum pressure in the model.
=======
APOLLO TO-DO LIST (SEPARATE FROM THE ISSUE LIST)
>>>>>>> Stashed changes

Set the cloud priors so that the cloud base cannot be deeper than the
    maximum pressure in the model.
Add an option to auto-detect the resolution of the input spectrum and
    sample the cross section files accordingly.
Convert the cross section tables to bin files and do direct lookup of the
    needed wavelengths.
Create lo-res opacity tables suitable to be added to the Github repository.
Add an option to plot comparison or literature values to Plot Apollo.
Re-evaluate specific mode spectrum generation. Would it be better to accept
    auxiliary input files like the photometric mode does instead of hard-coding
    all the JWST modes. Consider what specific mode options to include, if any
    since we don't have any other observatories. Clean up JWST mode file names,
    and add or delete appropriate files. Revise the AddNoise function based on
    the ETC or otherwise fix the inconsistencies.

*******************************************************************************

APOLLO CHANGE LOG

v0.13
Added additional useful Python packages.
Move several code elements to different files, including default parameter
values (through the defaults.py file), parameter names, and temperature-
pressure profile calculations.
Defined variables for Earth radius, Jupiter radius, and parsec.
Added a new 'P_profile' variable and P_points.py file, which allows the user to
select a non-uniformly-spaced pressure profile from a dictionary.
Added a new 'plotting' variable that allows the used to toggle plotting on and
off, which is often necessary to run the code on a cluster.
Converted some lists to arrays for easier calculations.
Added dummy derived variables to GetModel() in case they cannot be calculated.
Added the TP_functions and TP_profiles.py that allow the user to select a
different T-P profile parameterization from a dictionary.
Further refined the output parameter file format.
Moved computation of derived parameters to GetModel().
Clarified some error messages.
Fixed the bug that caused an error if lnprob() was initially called before
GetModel().
Fixed a bug in the initialization of medianparams.
Removed superfluous function arguments.
Changed the molecular weights and Rayleigh cross sections to vectors (which
helps with computing contribution functions).
Implemented a cloud fill fraction parameter.
Cleaned up the ConvSpec() function.
Set h2+he as the automatic default filler gas.
Revised the SliceModel() function for robustness.
Implemented a power law cloud model (Cloud Model 4).
Added a default case for no clouds to call get_ClearSpectrum() only to speed
execution.
Added functionality to Plot.Apollo.py to plot the parametric T-P profile.
Added commented-out code for radius corrections to Plot Apollo.
Improved appearance of corner plots in Plot Apollo.
Implemented a corner plot of statistical parameters in Plot Apollo.

v0.12.2
Significant changes to clean up handling of multiple bands, including separate
deltaL and logf parameters for each band.
Changed logf to actually start from the logarithm. (This fixed the problem
with feeding output files back in.)
Removed the alternate, ratio-based treatment of radius, keeping it as just Rad.
Removed rescaling between Earth radii and Jupiter radii, which was screwing up
the results if the radius was omitted.
Simplified Plot Apollo to read in relevant parameters from the input file. Also
changed the samples output file to match.
Fixed the "likelihood of input parameters" calculations (retrieval mode only).
Tweaked band handling for reliability.
Fixed the header of the Samples files when there are fixed parameters.
Fixed a bug that returned a divide-by-zero error in C/O when there was no
oxygen in the atmosphere.
Increased fixed the number of decimal places written for wavelengths from two
to five.
Changed the input format for mass to "Mass_Limits", minmass, maxmass.
Corrected formatting errors in the output parameter file.
Verified that plot generation runs in APOLLO main.

v0.12.1
Fixed a bug in the "no-radius" normalization.
Fixed several problems with the binning and convolving functions.
Changed BinBands to accept 2D arrays and return flattened arrays regardless of
the number of bands.
Converted the whole code from dealing with wavenumbers to dealing with
wavelengths because that's much less confusing.
Fixed the delta-L handing, which was mismatched in wavelength v. wavenumber.

AddNoise.py
import os
import numpy as np
import matplotlib.pyplot as plt
import Filter
import sys
from matplotlib import rc
from scipy import interpolate
rc('text',usetex=True)

def addNoise(mode,obstype,mod_wave,mod_flux,noise_params):
    # mod_flux = emission from the target in (erg s^-1 cm^-2 Hz^-1)
    
    # Fundamental constants
    G = 6.67259e-8
    c = 2.9979e10
    h = 6.626075e-27
    kB = 1.380758e-16
    sig = 5.67e-5
    u = 1.66054e-24
    Na = 6.02214e23
    pi = 3.14159265

    # Model Parameters
    Rp = noise_params[0]
    Rs = noise_params[1]
    dist = noise_params[2]
    RA = noise_params[3]
    Dec = noise_params[4]
    Teff = noise_params[5]
    duration = noise_params[6]    # exposure time in hours
    ntran = 1         # number of transits if applicable

    # compute solid angles of star and planet
    theta_star = 7.0e5*Rs/(3.0e13*dist)
    theta_planet = 7.0e4*Rp/(3.0e13*dist)
    omega_star = pi*theta_star*theta_star
    omega_planet = pi*theta_planet*theta_planet

    length2 = len(mod_wave)
    planet_flux = np.zeros(length2)
    bnu = np.zeros(length2)
    star_flux = np.zeros(length2)

    j=length2-1
    k=0

    uwave = mod_wave
    mod_wave = 10000./mod_wave

    for i in range(0,len(mod_flux)):
        planet_flux[i] = (float(mod_flux[i])*(0.5/pi))*omega_planet  # multiply by solid angle
        
    received_flux = planet_flux                  # actual flux density (erg/s/cm^2/Hz) at the telescope
    planet_flux = planet_flux/(h*c*uwave)        # divide by photon energy
    planet_flux = planet_flux*2.5e5              # multiply by collecting area
    # this should be received flux in \gamma s^-1 aperture^-1 Hz^-1

    bnu = 2.0*h*c*pow(uwave,3)/(np.exp(1.44*uwave/Teff)-1.0)

    star_flux = bnu*omega_star
    star_flux = star_flux/(h*c*uwave)
    star_flux = star_flux*2.5e5

    fil = Filter.Filter()
    
    if mode<=8:
        xthruput = fil.NIRSpec(mode)
        oname = 'NIRSpec4'
    if mode==9:
        xthruput = fil.MIRILRS(True)  # slit spectroscopy
        oname = 'MIRISlit'
    if mode==10:
        xthruput = fil.MIRILRS(False) # slitless spectroscopy
        oname = 'MIRISlitless'
    if mode==11:
        xthruput = fil.NIRCamF322W2()
        oname = 'NIRCam'
    if mode==12:
        xthruput = fil.NIRISS()
        oname = 'NIRISS'
    if mode==13:
        xthruput = fil.MIRIMRS('A')
        oname = 'MIRIA'
    if mode==14:
        xthruput = fil.MIRIMRS('B')
        oname = 'MIRIB'
    if mode==15:
        xthruput = fil.MIRIMRS('C')
        oname = 'MIRIC'

    # begin observe.pro -- this does not change the units -- \gamma s^-1 aperture^-1 Hz^-1
    num = int(fil.resolve)
    num6 = int(6.0*fil.resolve)
    sigma = num/2.35
    kernel = np.zeros(num6)
    for i in range(0,num6):
        kernel[i] = np.exp(-0.5*(i-num6/2.)*(i-num6/2.)/sigma/sigma)
    kernel = kernel/np.sum(kernel)
    flux_planet = np.convolve(planet_flux,kernel,mode='same')
    flux_density = np.convolve(received_flux,kernel,mode='same')
    flux_star = np.convolve(star_flux,kernel,mode='same')
    
    f = interpolate.interp1d(mod_wave,flux_star,kind='linear')
    flux_star = f(fil.calwave)
    f2 = interpolate.interp1d(mod_wave,flux_planet,kind='linear')
    flux_planet = f2(fil.calwave)
    f3 = interpolate.interp1d(mod_wave,flux_density,kind='linear')
    flux_density = f3(fil.calwave)
    
    # begin background.pro
    dum = len(fil.calwave)
    num = dum
    background = np.zeros(num)
    zodi = np.zeros(num)

    # implement standard method EULER to convert RA and DEC to ecliptic coordinates                  
    equinox = '(J2000)'
    psi = 0.0
    stheta = 0.39777715593
    ctheta = 0.91748206207
    phi = 0.0

    ao = RA/57.296 - phi
    bo = Dec/57.296
    sb = np.sin(bo)
    cb = np.cos(bo)
    cbsa = cb*np.sin(ao)
    bo = -stheta*cbsa + ctheta*sb
    elat = np.arcsin(np.modf(bo)[0])

    ao = np.arctan((ctheta*cbsa + stheta*sb) / cb*np.cos(ao)) # make sure output is in the right quadrant
    elong = np.fmod(ao+psi+4*pi,2*pi)
    # end EULER

    elat = np.fabs(elat)

    # localzodi constants
    tau_lz90 = 4.0e-8
    A_lz = 0.6
    alpha = -0.4
    T_lz = 265
    elatrad = elat/57.296

    for i in range(0,num):
        # implement localzodi
        l_cm = fil.calwave[i]*1.e-4
        B_nu = (2.0*h*c/pow(l_cm,3)) / (np.exp(h*c/kB/T_lz/l_cm)-1)
        B_lz = (tau_lz90 * B_nu) / np.sqrt( np.sin(elatrad)*np.sin(elatrad) + pow(A_lz*pow(fil.calwave[i]/11.0,alpha)*np.cos(elatrad),2))
        zodi[i] = B_lz*1.e17
        # end localzodi
        
        # implement JWST_back
        freq = 3.0e14/fil.calwave[i]
        hnk = 1.44e4/fil.calwave[i]
        const = 2.0*h*pow(freq,3)/9.0e20
        T1 = 38.3
        T2 = 44.5
        T3 = 92.0
        T4 = 155.0
        B1 = const/(np.exp(hnk/T1)-1.0)
        B2 = const/(np.exp(hnk/T2)-1.0)
        B3 = const/(np.exp(hnk/T3)-1.0)
        B4 = const/(np.exp(hnk/T4)-1.0)
        S = 0.48*B1 + 0.1*B2 + 3.0e-5*B3 + 9.9e-7*B4
        background[i] = 1.0e17*S
        # end JWST_back

    if(fil.slit<0.0):
        xback = np.sum(background)
        for i in range(0,len(background)): background[i] = xback
        zback = np.sum(zodi)
        for i in range(0,len(zodi)): zodi[i] = zback

    background = 1.0e-17*background
    zodi = 1.0e-17*zodi
    background = background/(h*c*(1.0e4/fil.calwave))
    zodi = zodi/(h*c/(fil.calwave/1.e4))
    # end background.pro

    if(fil.slit > 0.): omeg_back = (fil.slit*fil.scale/206265.)*(1.22*fil.calwave/6.0e6)
    if(fil.slit < 0.): omeg_back = (fil.scale/206265.)*(1.22*fil.calwave/6.0e6)

    if(np.min(fil.calwave) < 0.9): omeg_back = (fil.scale/206265.)*(22.0*fil.scale/206265.)

    background = background * omeg_back * 2.5e5
    zodi = zodi * omeg_back * 2.5e5

    freq1 = 1.e4/fil.calwave
    bandwidth = np.fabs(np.gradient(freq1)) * c

    if(np.max(fil.calwave) > 15.0):
        bandwidth[fil.npix1-1] = bandwidth[fil.npix1-2]
        bandwidth[fil.npix1] = bandwidth[fil.npix1+1]
        bandwidth[fil.npix1+fil.npix2-1] = bandwidth[fil.npix1+fil.npix2-2]
        bandwidth[fil.npix1+fil.npix2] = bandwidth[fil.npix1+fil.npix2+1]
        bandwidth[fil.npix1+fil.npix2+fil.npix3-1] = bandwidth[fil.npix1+fil.npix2+fil.npix3-2]
        bandwidth[fil.npix1+fil.npix2+fil.npix3] = bandwidth[fil.npix1+fil.npix2+fil.npix3+1]

    # Resolved/directly-imaged
    if obstype==0:
        flux_signal = flux_planet  # \gamma s^-1 aperture^-1 Hz^-1
        flux_noise = zodi
    # Secondary eclipse
    if obstype==1:
        flux_signal = flux_planet  # \gamma s^-1 aperture^-1 Hz^-1
        flux_noise = zodi + flux_star
    # Transit
    if obstype==2:
        flux_signal = flux_star  # \gamma s^-1 aperture^-1 Hz^-1
        flux_noise = zodi
        
    electron_signal = flux_signal * xthruput  # e- s^-1 Hz^-1
    electron_noise = flux_noise * xthruput
    
    electron_signal = electron_signal + background
    electron_noise = electron_noise + background

    electron_signal = bandwidth * electron_signal  # e- s^-1 bin^-1
    electron_noise = bandwidth * electron_noise
    flux_density = bandwidth * flux_density
    # end observe.pro

    signal_electrons = 3600.*ntran*duration*fil.effic*electron_signal  # e- exposure^-1 bin^-1
    noise_electrons = np.sqrt(3600.*ntran*duration*fil.effic*2.0*(electron_signal+electron_noise))
    
    real_signal = 3600.*ntran*duration*fil.effic*electron_signal
    real_noise = noise_electrons/real_signal
    fnoise = noise_electrons/signal_electrons

    return fil.calwave,flux_density,fnoise
